enclave/seal.c

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/enclave.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/internal/crypto/gcm.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/safecrt.h>
+#include <openenclave/internal/entropy.h>
+
+static const uint8_t _iv[12] = { 0 };
+
+oe_result_t oe_seal(
+    const uint8_t* key_info,
+    size_t key_info_size,
+    const uint8_t* plaintext,
+    size_t plaintext_size,
+    const uint8_t* additional_data,
+    size_t additional_data_size,
+    uint8_t **blob,
+    size_t *blob_size)
+{
+    uint8_t* key = NULL;
+    size_t key_size = 0;
+
+    oe_sealed_blob_header_t *header;
+    oe_entropy_kind_t k;
+    uint8_t* payload;
+    size_t size;
+    oe_result_t result = OE_OK;
+
+    if (key_info == NULL || key_info_size != sizeof(header->key_info) ||
+        blob == NULL || blob_size == NULL)
+        return OE_INVALID_PARAMETER;
+
+    size = sizeof(*header);
+    if (size > OE_UINT32_MAX - plaintext_size)
+        return OE_INVALID_PARAMETER;
+    size += plaintext_size;
+    if (size > OE_UINT32_MAX - additional_data_size)
+        return OE_INVALID_PARAMETER;
+    size += additional_data_size;
+
+    header = (oe_sealed_blob_header_t*)oe_calloc(1, size);
+    if (header == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    OE_CHECK(oe_memcpy_s(&header->key_info, sizeof(header->key_info),
+                         key_info, key_info_size));
+
+    OE_CHECK(oe_get_entropy(header->key_info.key_id,
```

Comment:
Why is this writing a new random value into the provided `key_info.key_id` value that should have been set up as the dev intended in `oe_get_seal_key_info` already?

Reply:
As mentioned earlier, the IV is constant so the key must be fresh. We could rely on the developer not to reuse keys but that'd be error prone, especially for those unfamiliar with GCM. Given `key_id` will be used only once and be embedded into the seal blob, there shall be no reason for developers to use specific values. Thus I added this extra code to generate fresh `key_id` on every invocation.